bionic/utils/reload.py

```diff
@@ -0,0 +1,40 @@
+import builtins
+from importlib import reload
+from sys import modules as module_registry
+
+
+def recursive_reload(module):
+    """
+    Helper method to reload a module recursively. If a module imports a set of
+    modules, then the modules in the set are also reloaded and so on.
+
+    Not all modules can be reloaded. This method attempts to reload as many
+    modules as possible.
+
+    Also note that this method may not be able to handle dynamic imports that
+    only happens at runtime. For example, if a module imports another module
+    only when a certain method is executed, reloading the former module does
+    not guarantee that the latter module is reloaded.
+    """
+
+    original_import = builtins.__import__
+    already_reloaded = set()
+
+    def custom_import(name, globals=None, locals=None, fromlist=[], level=0):
```

Comment:
This will re-import all the third-party libraries and so on that the root module uses, right? That seems like it could be quite a lot. Do you think it could make sense to avoid reloading things like standard libraries and installed libraries?

Reply:
It definitely makes sense. The problem is figuring out which modules are standard libraries and which ones are not. It would be nice if there is a simple heuristics to do this. Right now, there is a `hasattr(module, "__file__")` check that avoids reloading modules that does not contain python source code.

Maybe we can check file timestamp?

Reply:
Could we figure this out from the location of the module files? I think a human being could generally do this by examining the file paths, but I'm not sure exactly what the algorithm should be.

The file timestamp idea is also interesting; that does seem like a good way to filter out modules that don't need to be loaded.

Reply:
BTW, @namanjain do you know if Streamlit has any ideas that might be useful here?

Reply:
Streamlit uses the same logic for reloading as caching. They check that the file should be in the dir where the script is running from or is in PYTHONPATH, and the file should not be blacklisted. This doc is not publicly available, but you can see this logic for caching in the Smart Caching doc [here](https://docs.google.com/document/d/1VxF-GC-Fk8JC-yTr1ACrAIjKBdevsH5sAnNp9DN6nA0/edit#heading=h.7obdz6b2lnx2). Simon and I briefly discussed that we should share that logic between smart caching and reloading, but we never reached a conclusion. I think it would be pretty nice to share that logic though.

You can find how Streamlit filters it [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/watcher/local_sources_watcher.py#L154).

Reply:
Thanks! Checking PYTHONPATH should work.

Reply:
Using streamlit's `file_in_pythonpath` method to skip over installed modules.
`file_in_pythonpath` uses environment variables and stays constant even if `sys.path` is modified. (The tests modify `sys.path`)

Reply:
I'm confused. Installed libraries generally don't go in the PYTHONPATH; in fact, it's normal to run Python without a PYTHONPATH at all. And Streamlit seems to use PYTHONPATH the opposite way: it reloads files that *are* in the PYTHONPATH, right?

Reply:
Ah, I think I have misunderstood how `PYTHONPATH` works. Running python or pytest from command line shows that `PYTHONPATH` can be empty.

Reply:
I have reverted the changes. How about letting this method reload every module for now, while we figure out a solution on how to reduce the reload separately? Reloading everything is the simplest approach, is guaranteed to work and solves the main problem.

Reply:
That's true, but it also feels like a bit of a regression from the previous behavior. Is there something simple we could do, like not reload files that are within a separate `site-packages` directory? That seems like it might catch most installed libraries.

Reply:
Added the sysconfig-based solution that I have been looking into. Also added a test to check that internal modules are not reloaded.